source/rsa.c

```diff
@@ -0,0 +1,285 @@
+    AWS_PRECONDITION(key_pair);
+    AWS_PRECONDITION(out);
+
+    if (AWS_UNLIKELY(aws_rsa_key_pair_max_encrypt_plaintext_size(key_pair, algorithm) < plaintext.len)) {
+        AWS_LOGF_ERROR(AWS_LS_CAL_RSA, "Unexpected buffer size. For RSA, ciphertext  is expected to match block size");
+        return aws_raise_error(AWS_ERROR_CAL_BUFFER_TOO_LARGE_FOR_ALGORITHM);
+    }
+
+    if (key_pair->good) {
+        return key_pair->vtable->encrypt(key_pair, algorithm, plaintext, out);
+    }
+
+    return aws_raise_error(AWS_ERROR_INVALID_STATE);
+}
+
+AWS_CAL_API int aws_rsa_key_pair_decrypt(
+    struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_encryption_algorithm algorithm,
+    struct aws_byte_cursor ciphertext,
+    struct aws_byte_buf *out) {
+    AWS_PRECONDITION(key_pair);
+    AWS_PRECONDITION(out);
+
+    if (AWS_UNLIKELY(ciphertext.len != (key_pair->key_size_in_bits / 8))) {
+        AWS_LOGF_ERROR(AWS_LS_CAL_RSA, "Unexpected buffer size. For RSA, ciphertext is expected to match block size.");
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
+    if (key_pair->good) {
+        return key_pair->vtable->decrypt(key_pair, algorithm, ciphertext, out);
+    }
+
+    return aws_raise_error(AWS_ERROR_INVALID_STATE);
+}
+
+int aws_rsa_key_pair_sign_message(
+    struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_signing_algorithm algorithm,
+    struct aws_byte_cursor digest,
+    struct aws_byte_buf *out) {
+    AWS_PRECONDITION(key_pair);
+    AWS_PRECONDITION(out);
+
+    if (key_pair->good) {
+        return key_pair->vtable->sign(key_pair, algorithm, digest, out);
+    }
+
+    return aws_raise_error(AWS_ERROR_INVALID_STATE);
+}
+
+int aws_rsa_key_pair_verify_signature(
+    struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_signing_algorithm algorithm,
+    struct aws_byte_cursor digest,
+    struct aws_byte_cursor signature) {
+    AWS_PRECONDITION(key_pair);
+
+    if (key_pair->good) {
+        return key_pair->vtable->verify(key_pair, algorithm, digest, signature);
+    }
+
+    return aws_raise_error(AWS_ERROR_INVALID_STATE);
+}
+
+size_t aws_rsa_key_pair_block_length(struct aws_rsa_key_pair *key_pair) {
+    AWS_PRECONDITION(key_pair);
+    AWS_PRECONDITION(key_pair->good);
+    return key_pair->key_size_in_bits / 8;
+}
+
+size_t aws_rsa_key_pair_signature_length(struct aws_rsa_key_pair *key_pair) {
+    AWS_PRECONDITION(key_pair);
+    AWS_PRECONDITION(key_pair->good);
+    return key_pair->key_size_in_bits / 8;
+}
+
+int aws_rsa_key_pair_get_public_key(
+    const struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_key_export_format format,
+    struct aws_byte_cursor *out) {
+    (void)format; /* ignore format for now, since only pkcs1 is supported. */
+    AWS_PRECONDITION(key_pair);
+    AWS_PRECONDITION(out);
+
+    if (key_pair->good) {
+        *out = aws_byte_cursor_from_buf(&key_pair->pub);
+        return AWS_OP_SUCCESS;
+    }
+
+    return aws_raise_error(AWS_ERROR_INVALID_STATE);
```

Comment:
Function declaration mentions that this isn't supported on Windows. We probably want AWS_ERROR_PLATFORM_NOT_SUPPORTED raised when that is the case